### PR TITLE
Extensions.Hosting improvements.

### PR DESCRIPTION
**Pull request by cijothomas**
services.AddOpenTelemetryTracing() now looks for any IConfigureOptions<TracerProviderBuilder>, and invokes them one by one.
See example usage in the Test `AddOpenTelemetryTracingE2EDemo` in this PR

TODO: Find out how to obtain `IServiceProvider` in the Configure<TracerProviderBuilder>.

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs [hunk 1]
```diff
@@ -76,44 +81,78 @@ public void AddOpenTelemetryTracerProvider_HostBuilt_OpenTelemetrySdk_Registered
...
+            var testProcessor = new TestActivityProcessor();
+            var resource = Resources.Resources.CreateServiceResource("serviceName");
+            testProcessor.StartAction = (act) => Assert.Equal("MyActivityName", act.DisplayName);
+            testProcessor.EndAction = (act) => Assert.Equal(resource, act.GetResource());
+            var builder = new HostBuilder().ConfigureServices(services =>
             {
-                builder.AddInstrumentation(() => provider.GetRequiredService<TestInstrumentation>());
-            }));
+                // Add OpentelemetryTracing.
+                services.AddOpenTelemetryTracing();
```
Comment (cijothomas) [inline test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs:126]: just to show - no impact as only registered once.

File: test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs [hunk 2]
```diff
@@ -76,44 +81,78 @@ public void AddOpenTelemetryTracerProvider_HostBuilt_OpenTelemetrySdk_Registered
...
+                services.Configure<TracerProviderBuilder>((builder) => builder.SetSampler(new AlwaysOnSampler()));
+
+                // Keep configuring TracerProviderBuilder
+                services.Configure<TracerProviderBuilder>((builder) => builder.AddProcessor(testProcessor));
+
+                // Keep configuring TracerProviderBuilder
+                services.Configure<TracerProviderBuilder>((builder) => builder.SetResource(resource));
+
+                // Keep configuring TracerProviderBuilder
+                services.Configure<TracerProviderBuilder>((builder) => builder.AddSource(TestActivitySourceName));
```
Comment (cijothomas) [inline test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs:140]: to demo - how to build the TracerProviderBuilder from any place. The order in which Configure is called is the order in which they are called.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs [hunk 3]
```diff
@@ -35,110 +37,37 @@ public static class OpenTelemetryServicesExtensions
...
-        /// <summary>
-        /// Adds OpenTelemetry TracerProvider to the specified <see cref="IServiceCollection" />.
-        /// </summary>
-        /// <param name="services">The <see cref="IServiceCollection" /> to add services to.</param>
-        /// <param name="configure">The <see cref="TracerProviderBuilder"/> action to configure TracerProviderBuilder.</param>
-        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
-        public static IServiceCollection AddOpenTelemetryTracing(this IServiceCollection services, Action<TracerProviderBuilder> configure)
-        {
-            if (configure is null)
+            services.AddSingleton<TracerProvider>((s) =>
```
Comment (reyang) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:40]: Stepping back question - for a console application it is possible to use multiple `TracerProvider` instances, with each provider having its own resource/instrumentation/sampler/processor.
Do we see a scenario here, or we always expect a singleton provider?

File: test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs [hunk 4]
```diff
@@ -76,44 +81,78 @@ public void AddOpenTelemetryTracerProvider_HostBuilt_OpenTelemetrySdk_Registered
...
-
             Assert.Same(tracerProvider1, tracerProvider2);
         }
 
         [Fact]
-        public void AddOpenTelemetryTracerProvider_ServiceProviderArgument_ServicesRegistered()
+        public async Task AddOpenTelemetryTracingDetectMultipleCalls()
         {
             var testInstrumentation = new TestInstrumentation();
+            int instrumentationCallBackCount = 0;
```
Comment (reyang) [inline test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs:91]: Nit - Callback is a single word https://en.wikipedia.org/wiki/Callback_(computer_programming).
```suggestion
            int instrumentationCallBackCount = 0;
```
```suggestion
            int instrumentationCallbackCount = 0;
```

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs [hunk 5]
```diff
@@ -35,110 +37,37 @@ public static class OpenTelemetryServicesExtensions
         /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
         public static IServiceCollection AddOpenTelemetryTracing(this IServiceCollection services)
```
Comment (reyang) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:38]: Curious about the naming convention - do we want `AddOpenTelemetryTracing` or `AddOpenTelemetryTracerProvider`?

File: test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs [hunk 6]
```diff
@@ -125,5 +164,54 @@ public void Dispose()
                 this.Disposed = true;
             }
         }
+
+        internal class TestActivityProcessor : BaseProcessor<Activity>
```
Comment (reyang) [inline test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs:168]: Consider reusing the `./test/OpenTelemetry.Tests/Shared/TestActivityProcessor.cs`.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:40]: If the Asp.Net Core (Or DI based ) requires more than one TracerProvider, they can still use the non DI method to create that. Don't think we need to support that using the DI extension methods. I don't see a need yet for that.

File: test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs (hunk 6 above)
Comment (cijothomas) [reply test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs:168]: yes. thats the plan.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs (hunk 5 above)
Comment (cijothomas) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:38]: In a contested mid-year election, `AddOpenTelemetryTracing` won with huge margin against `AddOpenTelemetryTracerProvider` as the later was too lengthy
https://github.com/open-telemetry/opentelemetry-dotnet/issues/1196

**Review by reyang — APPROVED** [review]
LGTM.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs [hunk 7]
```diff
@@ -35,110 +37,37 @@ public static class OpenTelemetryServicesExtensions
...
+                    config.Configure(builder);
+                }
 
-        /// <summary>
-        /// Adds OpenTelemetry TracerProvider to the specified <see cref="IServiceCollection" />.
-        /// </summary>
-        /// <param name="services">The <see cref="IServiceCollection" /> to add services to.</param>
-        /// <param name="configure">The <see cref="TracerProviderBuilder"/> action to configure TracerProviderBuilder.</param>
-        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
-        public static IServiceCollection AddOpenTelemetryTracing(this IServiceCollection services, Action<IServiceProvider, TracerProviderBuilder> configure)
```
Comment (alanwest) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:67]: New Relic's exporter uses `ILogger` for diagnostic logging. We use this overload that accepts an `IServiceProvider` in our examples to demonstrate how to configure our exporter with logging enabled. Is there a better way to achieve this?

See: https://github.com/newrelic/newrelic-telemetry-sdk-dotnet/blob/281223149ad73ac94f87fd7eafc41fa278bdbb1d/examples/NewRelic.OpenTelemetry/AspNetCore/Startup.cs#L25-L40

File: test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs [hunk 8]
```diff
@@ -76,44 +81,78 @@ public void AddOpenTelemetryTracerProvider_HostBuilt_OpenTelemetrySdk_Registered
...
-            services.AddSingleton(testInstrumentation);
-            services.AddOpenTelemetryTracing((provider, builder) =>
+            var builder = new HostBuilder().ConfigureServices(services =>
             {
-                builder.AddInstrumentation(() => provider.GetRequiredService<TestInstrumentation>());
+                services.AddOpenTelemetryTracing();
+                services.AddOpenTelemetryTracing();
+                services.AddOpenTelemetryTracing();
+                services.AddOpenTelemetryTracing();
+                services.Configure<TracerProviderBuilder>((builder) =>
```
Comment (tekian) [inline test/OpenTelemetry.Extensions.Hosting.Tests/HostingExtensionsTests.cs:99]: Would it be possible to use `services.AddOpenTelemetryTracing((builder) => { ... })` even for extensions rather then `services.Configure<TracerProviderBuilder>((builder) => {...})`. I presume the tests are to demonstrate how a consumer of the library would use it. And if that's so, using `services.Configure<T>` to mutate TracerProviderBuilder seems like we are leaking implementation detail. (We can still use the same mechanics behind the scenes).

**Comment by cijothomas** [comment]
Closing this PR as this is not actively being worked on. Will be back to solving the same issue (there are some attempts already)

**Pull request closed without merging**